src/blockMonitor/blockCache.ts

```diff
@@ -0,0 +1,230 @@
+            // If the block's parent is above the minimum visible height, it needs to be added first
+            if (!this.canAddBlock(block)) {
+                throw new ApplicationError("Tried to add a block before its parent block.");
+            }
+        }
+
+        // Update data structures
+
+        // Save block stub
+        const newBlockStub = this.makeBlockStub(block.hash, block.number, block.parentHash);
+        this.blockStubsByHash.set(block.hash, newBlockStub);
+
+        // Add set of transactions for this block hash
+        this.txHashesByBlockHash.set(block.hash, new Set(block.transactions));
+
+        // Index block by its height
+        addItemToKeyedSet(this.blockHashesByHeight, block.number, block.hash);
+
+        // Link all the tx hashes of this block to the block hash
+        for (let txHash of block.transactions || []) {
+            // Add this block among the block containing transaction txHash
+            addItemToKeyedSet(this.blockHashesByTxHash, txHash, block.hash);
+        }
+
+        // If the maximum block height increased, we might have to prune some old info
+        if (this.mMaxHeight < block.number) {
+            this.mMaxHeight = Math.max(this.mMaxHeight, block.number);
+            this.prune();
+        }
+
+        return true;
+    }
+
+    /**
+     * Returns the `BlockStubChain` for the block with hash `blockHash`, or `null` if the block is not in cache.
+     * @param blockHash
+     */
+    public getBlockStubChain(blockHash: string): BlockStubChain | null {
+        const blockStubChain = this.blockStubsByHash.get(blockHash);
+        if (blockStubChain === undefined) {
+            return null;
+        }
+        return blockStubChain;
+    }
+
+    /**
+     * Returns the `IBlockStub` for the block with hash `blockHash`, or `null` if the block is not in cache.
+     * @param blockHash
+     */
+    public getBlockStub(blockHash: string): IBlockStub | null {
+        const blockStubChain = this.getBlockStubChain(blockHash);
+        if (blockStubChain === null) {
+            return null;
+        }
+        return blockStubChain.asBlockStub();
+    }
+
+    /**
+     * Returns true if the block with hash `blockHash` is currently in cache.
+     **/
+    public hasBlock(blockHash: string) {
+        return this.blockStubsByHash.has(blockHash);
+    }
+
+    /**
+     * Returns number of confirmations using `headBlockHash` as tip of the blockchain, looking for `txHash` among the ancestor blocks;
+     * return 0 if no ancestor containing the transaction is found.
+     * Note: This will return 0 for transactions already at depth bigger than `this.maxDepth` when this function is called.
+     */
+    public getConfirmations(headBlockHash: string, txHash: string): number {
+        let depth = 0;
+        let curBlock = this.getBlockStub(headBlockHash);
+        while (curBlock !== null) {
+            const txsInCurBlock = this.txHashesByBlockHash.get(curBlock.hash) || new Set();
```

Comment:
very minor point, but instead of creating a new set you could check for null:
```
if (txsInCurBlock && txsInCurBlock.has(txHash)) {
                return depth + 1;
            }
```

Reply:
Done.